lib/opentelemetry.js

```diff
@@ -0,0 +1,186 @@
+// Initialize OpenTelemetry tracer singleton and export
+// the most useful API members, the tracer and some helper functions.
+//
+// The library shouldn't be used directly (although it's not fatal),
+// it is registered as self.apos.telemetry pseudo module (see index.js).
+const api = require('@opentelemetry/api');
+const version = require('../package.json').version;
+
+/** @type {api.Tracer} */
+let tracer;
+
+const AposAttributes = {
+  SCENE: 'apos.scene',
+  TEMPLATE: 'apos.template',
+  EVENT_MODULE: 'apos.event.module',
+  EVENT_NAME: 'apos.event.name',
+  // The module and method targeted by the current operation (see doc module)
+  TARGET_NAMESPACE: 'apos.target.namespace',
+  TARGET_FUNCTION: 'apos.target.function'
+};
+
+module.exports = function (options = {}) {
+  if (!tracer) {
+    tracer = api.trace.getTracer('apostrophe', version);
+
+    // This shouldn't be needed, but having it doesn't hurt
+    if (options.openTelemetryProvider) {
+      api.trace.setGlobalTracerProvider(options.openTelemetryProvider);
+    }
+  }
+
+  /**
+   * Start and return a new span. Optionally provide a parent span or allow the parent
+   * span to be auto-detected.
+   * Use this when the current code block does the tracing and it doesn't expect
+   * more tracing to happen down the line.
+   *
+   * @param {String} name span name
+   * @param {api.Span|Boolean} [parentSpan] optional parent span
+   * @param {api.SpanOptions} [options] span options
+   * @returns {api.Span}
+   * @example
+   * // Auto-merge with the currently active span context
+   * const span = self.apos.telemetry.aposStartSpan('event:someEvent');
+   * // Provide a parent span
+   * const span = self.apos.telemetry.aposStartSpan('event:someEvent', parent);
+   * // Do not merge with any parent, start as a root
+   * const span = self.apos.telemetry.aposStartSpan('event:someEvent', false);
+   * // Provide options for the newly created span
+   * const span = self.apos.telemetry.aposStartSpan('event:someEvent', true, {
+   *   attributes: [
+   *     [SemanticAttributes.HTTP_METHOD]: "GET",
+   *     [SemanticAttributes.HTTP_FLAVOR]: "1.1",
+   *     [SemanticAttributes.HTTP_URL]: req.url
+   *   ]
+   * });
+   */
+  function aposStartSpan(name, parentSpan, options) {
```

Comment:
I am still trying to understand the need for this. Is it because the root span for the startup process is always out there as a default "parent" and that messes up the tracing for regular requests if we rely on the normal autodetection of parents via AsyncLocalStorage that happens in OpenTelemetry?

Would that issue go away more simply if we ended the root span at the end of the initialization process?

Reply:
This is just fixing the not so good public API (IMO) of OTel. It gets rid of tons of boilerplate code from the core and makes it more readable.